File: Week11_BFS/symmetric-tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

class Solution:
    def checkSame(self, p: TreeNode, q: TreeNode) -> bool:
        if not p and not q:
            return True
        
        if not q or not p:
            return False
        
        if q.val == p.val:
            return True
        
        return False
    
    def isSymmetric(self, root: TreeNode) -> bool:
        queue = []
        queue.append(root.left)
        queue.append(root.right)
        
        while len(queue) != 0:
            nodeP = queue.pop(0)
            nodeQ = queue.pop(0)
            
            if not self.checkSame(nodeP, nodeQ):
                return False
            
            if nodeP:
                queue.append(nodeP.right)
                queue.append(nodeQ.left)
                queue.append(nodeP.left)
                queue.append(nodeQ.right)
            
        return True

File: Week11_BFS/symmetric-tree.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */

class Solution {
    public boolean checkSame(TreeNode p, TreeNode q) {
        if (p == null && q == null) return true;
        
        if (p == null || q == null) return false;
        
        if (q.val == p.val) return true;
        
        return false;
    }
    
    public boolean isSymmetric(TreeNode root) {
        Queue<TreeNode> queue = new LinkedList<>();
        queue.offer(root.right);
        queue.offer(root.left);
        
        while (!(queue.isEmpty())) {
            System.out.println(queue);
            
            TreeNode nodeP = queue.poll();
            TreeNode nodeQ = queue.poll();
            
            if (!(checkSame(nodeP, nodeQ))) return false;
            
            if (nodeP != null) {
                queue.offer(nodeP.left);
                queue.offer(nodeQ.right);
                queue.offer(nodeP.right);
                queue.offer(nodeQ.left);
            }
        }
        return true;
    }
}